gettransaction: add an argument to decode the transaction

darosior opened the pull request:
This PR adds a new parameter to the `gettransaction` call : `decode`. If set to `true`, it will add a new `decoded` field to the response. This mimics the behavior of `getrawtransaction`'s `verbose` argument to avoid using 2 calls if we want to decode a wallet transaction (`gettransaction` then `decoderawtransaction`).

Fix #16181 .

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1725,10 +1728,15 @@ static UniValue gettransaction(const JSONRPCRequest& request)
 
     isminefilter filter = ISMINE_SPENDABLE;
-    if(!request.params[1].isNull())
+    if (!request.params[1].isNull())
```
Unrelated change, remove?

darosior replied:
Thought it would be great to fix this nit while passing through, will remove this change

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1725,10 +1728,15 @@ static UniValue gettransaction(const JSONRPCRequest& request)
 
+    bool decode_tx = false;
+    if (!request.params[2].isNull())
+        if (request.params[2].get_bool())
```
`decode_tx = request.params[2].get_bool()`

promag reviewed (commented):
Concept ACK, please include tests and a release note.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1664,13 +1664,14 @@ static UniValue gettransaction(const JSONRPCRequest& request)
                 {
                     {"txid", RPCArg::Type::STR, RPCArg::Optional::NO, "The transaction id"},
                     {"include_watchonly", RPCArg::Type::BOOL, /* default */ "false", "Whether to include watch-only addresses in balance calculation and details[]"},
+                    {"decode", RPCArg::Type::BOOL, /* default */ "true", "Whether to add a field with the decoded transaction"},
```
Default to false?

darosior replied:
Yes, corrected.

darosior committed:
gettransaction: add an argument to decode the transaction

This adds a new boolean parameter 'decode' to the gettransaction call, which, if set to true, add a 'decoded' field to the result containing the decoded transaction

darosior commented:
@promag thank you for the review. I added the changes and will include the tests and release note soon.

darosior committed:
tests: Add a new functional test for gettransaction

darosior committed:
doc: Add release note for the new gettransaction argument

darosior commented:
> Concept ACK, please include tests and a release note.

Since there was no functional test for the `gettransaction` I created a new one. However it only tests the `decoded` field since I don't know if testing other would not be overkill.

promag commented on test/functional/rpc_gettransaction.py:
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env python3
```
Maybe add the test in test/functional/wallet_basic.p instead?

darosior replied:
Ok.

promag commented on test/functional/rpc_gettransaction.py:
```diff
@@ -0,0 +1,31 @@
+        self.nodes[0].generate(101)
+        txid = self.nodes[0].sendtoaddress(self.nodes[0].getnewaddress(), 1)
+        # False for 'include_watchonly', True for 'decode'
+        tx = self.nodes[0].gettransaction(txid, False, True)
```
You can use named arguments and drop the comment above (untested):
```py
tx = self.nodes[0].gettransaction(txid=txid, decode=True)
```

darosior replied:
Yes, tested it it works thanks :D

darosior commented:
Moved the test to [`test/functional/wallet_basic.py`](https://github.com/bitcoin/bitcoin/blob/master/test/functional/wallet_basic.py) and used named arguments.

fanquake commented on test/functional/test_runner.py:
```diff
@@ -198,6 +198,7 @@
     'feature_blocksdir.py',
     'feature_config_args.py',
     'rpc_help.py',
+    'rpc_gettransaction.py',
```
Please remove this entry now that the test is in `wallet_basic.py`.

darosior commented:
Rebased.

darosior commented:
Rebased.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1670,11 +1671,13 @@ static UniValue gettransaction(const JSONRPCRequest& request)
             "    ,...\n"
             "  ],\n"
             "  \"hex\" : \"data\"         (string) Raw data for transaction\n"
+            "  \"decoded\" : \"transaction\"         (json object) Optional, the decoded transaction\n"
```
Remove `\"transaction\"` as the value is not a string. Maybe mention that it's the same result as `decoderawtransaction`?

darosior replied:
done

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1693,6 +1696,10 @@ static UniValue gettransaction(const JSONRPCRequest& request)
         if(request.params[1].get_bool())
             filter = filter | ISMINE_WATCH_ONLY;
 
+    bool decode_tx = false;
```
nit, `const bool decode = request.params[2].isNull() ? false : request.params[2].get_bool();`

darosior replied:
done

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1693,6 +1696,10 @@ static UniValue gettransaction(const JSONRPCRequest& request)
             filter = filter | ISMINE_WATCH_ONLY;
 
+    bool decode_tx = false;
+    if (!request.params[2].isNull())
```
nit, otherwise add `{ }`.

promag commented on test/functional/wallet_basic.py:
```diff
@@ -496,6 +496,13 @@ def run_test(self):
         assert_equal(self.nodes[0].getaddressinfo(change)['ischange'], False)
 
+        # Test "decoded" field value in gettransaction response
+        self.log.info("Testing gettransaction decoding...")
```
nit, remove log.

darosior replied:
Is there a rationale for removing the log ? I added it to mimic the other tests' behavior

promag commented on test/functional/wallet_basic.py:
```diff
@@ -496,6 +496,13 @@ def run_test(self):
+        # Test "decoded" field value in gettransaction response
+        self.log.info("Testing gettransaction decoding...")
+        self.nodes[0].generate(1)
+        txid = self.nodes[0].sendtoaddress(self.nodes[0].getnewaddress(), 1)
```
nit, remove and use the last transaction below?

darosior replied:
done

promag commented on test/functional/wallet_basic.py:
```diff
@@ -496,6 +496,13 @@ def run_test(self):
+        self.nodes[0].generate(1)
+        txid = self.nodes[0].sendtoaddress(self.nodes[0].getnewaddress(), 1)
+        tx = self.nodes[0].gettransaction(txid=txid, decode=True)
+        assert_equal(tx["decoded"], self.nodes[0].decoderawtransaction(tx["hex"]))
```
Or you can simply change L488:
```diff
-tx = self.nodes[0].decoderawtransaction(self.nodes[0].gettransaction(txid)['hex'])
+tx = self.nodes[0].gettransaction(txid, decode=True)['decoded']
```

darosior replied:
Probably better to leave it as a separated test

darosior commented:
Some minor corrections according to last review.

darosior commented:
Some minor corrections according to last review.

promag reviewed (commented):
You could use https://github.com/bitcoin/bitcoin/commit/0b6ba66238c377116bc6c21e19cffbf1b6dfc788#diff-0f596a69ccfb4d07c007cd6fc993898dR50 (not yet in master) to do:
```py
assert_no_key('decoded', self.nodes[0].gettransaction(txid, decode=False))
```

jb55 commented on test/functional/wallet_basic.py:
```diff
@@ -496,6 +496,11 @@ def run_test(self):
         self.nodes[0].setlabel(change, 'foobar')
         assert_equal(self.nodes[0].getaddressinfo(change)['ischange'], False)
 
+        # Test "decoded" field value in gettransaction response
```
this is a weird nit, but just curious why there's a "Non-breaking space" character between the `#` and `Test` (utf-8 0xC2A0)

```
git cat-file -p d5ab374c48a7f2708acc5e122044a3a9cfc1ac6a | grep '#.Test "decoded"' | xxd | head -n1
```

darosior replied:
I think it's caused by a vim module (deoplete-jedi on this one I think)

jb55 commented on doc/release-notes.md:
```diff
@@ -99,6 +99,10 @@ Low-level Changes section below.
   command line arguments.
 
+* The `gettransaction` RPC now accepts a third (boolean) argument `decode`. If
+  set to `true`, a new `decoded` field will be added to the response containing
```
There's another one of those chars here after `decoded` and before `field`. Not a big deal, just was wondering what editor you are using that would be adding these randomly?

```
git cat-file -p a2b23686c65d7e8c7e69b7547322d2eea1cf3b44 | xxd | grep c2a0
00000f90: 6465 6460 c2a0 6669 656c 6420 7769 6c6c  ded`..field will
```

darosior replied:
Vim, you right it's weird

jb55 reviewed (commented):
Concept ACK

darosior commented:
Rebased.
*(And addressed the spacing nits pointed by @jb55 :-) )*

achow101 commented on doc/release-notes-16185.md:
```diff
@@ -0,0 +1,3 @@
+RPC changes
+-----------
+The `gettransaction` RPC now accepts a third (boolean) argument `decode`. If set to `true`, a new `decoded` field will be added to the response containing
```
I think the sentence got cut off here.

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1724,6 +1729,12 @@ static UniValue gettransaction(const JSONRPCRequest& request)
 
+    if (decode_tx) {
+        UniValue decoded(UniValue::VOBJ);
+        TxToUniv(*wtx.tx, uint256(), decoded, false, pwallet->chain().rpcSerializationFlags());
```
nit: It's not necessary to have `pwallet->chain().rpcSerializationFlags()`.

achow101 reviewed (commented):
Concept ACK

darosior commented:
@achow101 thank you for the review, force pushed to rectify.

achow101 commented:
ACK a636daba4f61c32ec83d6677b8cff66dbde3fda0 Reviewed code and tested locally

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1676,11 +1677,13 @@ static UniValue gettransaction(const JSONRPCRequest& request)
             "    ,...\n"
             "  ],\n"
             "  \"hex\" : \"data\"         (string) Raw data for transaction\n"
+            "  \"decoded\" : transaction         (json object) Optional, the decoded transaction\n"
```
This seems wanting... but expanding here would be ugly.

luke-jr reviewed (approved):
utACK

meshcollider reviewed (commented):
utACK a636daba4f61c32ec83d6677b8cff66dbde3fda0

Once rebased, I will merge this

darosior commented:
Rebased.

promag commented:
You must make linters happy.

darosior commented:
> You must make linters happy.

No that's a Travis error : `curl: (7) Failed to connect to storage.googleapis.com port 443: Connection timed out`

meshcollider reviewed (commented):
re-utACK 9965940e35c445ccded55510348af228ff22f0e9

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1684,11 +1685,13 @@ static UniValue gettransaction(const JSONRPCRequest& request)
             "    ,...\n"
             "  ],\n"
             "  \"hex\" : \"data\"         (string) Raw data for transaction\n"
+            "  \"decoded\" : transaction         (json object) Optional, the decoded transaction\n"
```
Should mention that the field is identical to `decoderawtransaction`?

jnewbery commented:
Sorry for the post-merge review comment, but would it make sense to call the new parameter `verbose`, to mirror the `verbose` argument in `getrawtransaction`?

If so, we should make that change before v0.19 to avoid an API change.

ryanofsky commented:
> Sorry for the post-merge review comment, but would it make sense to call the new parameter `verbose`, to mirror the `verbose` argument in `getrawtransaction`?
> 
> If so, we should make that change before v0.19 to avoid an API change.

This seems like a good idea. You could open a one-line PR and add the milestone to get more feedback.

promag commented:
+1 for API consistency.
